source/tutorials/packaging-existing-software.md

```diff
@@ -372,30 +428,97 @@ error: builder for '/nix/store/bw2d4rp2k1l5rg49hds199ma2mz36x47-icat.drv' failed
+:::{tip}
+The easiest way to find what you need is on search.nixos.org/packages.
 :::
 
+### search.nixos.org
+
+Unfortunately in this case, [searching for `x11`](https://search.nixos.org/packages?query=x11) produces too many irrelevant results because X11 is ubiquitous.
+On the left side bar there is a list package sets, and [selecting `xorg`](https://search.nixos.org/packages?channel=23.11&buckets={%22package_attr_set%22%3A[%22xorg%22]%2C%22package_license_set%22%3A[]%2C%22package_maintainers_set%22%3A[]%2C%22package_platforms%22%3A[]}&query=x11) shows something promising.
+
+In case all else fails, it helps to become familiar with searching the [Nixpkgs source code](https://github.com/nixos/nixpkgs) for keywords.
+
+### Git and `rg`
+
+To find name assignments in the source, search for `"<keyword> ="`.
+For example, these are the search results for [`"x11 = "`](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22x11+%3D%22&type=code) or [`"libx11 ="`](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22libx11+%3D%22&type=code) on Github .
+
+Or fetch a local clone of the repository and use `rg`.
+Nixpkgs is huge.
+Only clone the latest revision if you don't want to wait a long time:
+
+```console
+$ nix-shell -p git ripgrep
+[nix-shell:~]$ git glone https://github.com/NixOS/nixpkgs --depth 1
+```
+Nixpkgs is *huge*, therefore specify which subdirectory you want to search:
+
+```console
+[nix-shell:~]$ rg "x11 =" pkgs
+pkgs/tools/X11/primus/default.nix
+21:  primus = if useNvidia then primusLib_ else primusLib_.override { nvidia_x11 = null; };
+22:  primus_i686 = if useNvidia then primusLib_i686_ else primusLib_i686_.override { nvidia_x11 = null; };
+
+pkgs/applications/graphics/imv/default.nix
+38:    x11 = [ libGLU xorg.libxcb xorg.libX11 ];
+
+pkgs/tools/X11/primus/lib.nix
+14:    if nvidia_x11 == null then libGL
+
+pkgs/top-level/linux-kernels.nix
+573:    ati_drivers_x11 = throw "ati drivers are no longer supported by any kernel >=4.1"; # added 2021-05-18;
+... <a lot more results>
+```
+
+Since `rg` is case sensitive by default,
+Add `-i` to make sure you don't miss anything:
+
+```
+[nix-shell:~]$ rg -i "libx11 =" pkgs
+pkgs/applications/version-management/monotone-viz/graphviz-2.0.nix
+55:    ++ lib.optional (libX11 == null) "--without-x";
+
+pkgs/top-level/all-packages.nix
+14191:    libX11 = xorg.libX11;
+
+pkgs/servers/x11/xorg/default.nix
+1119:  libX11 = callPackage ({ stdenv, pkg-config, fetchurl, xorgproto, libpthreadstubs, libxcb, xtrans, testers }: stdenv.mkDerivation (finalAttrs: {
+
+pkgs/servers/x11/xorg/overrides.nix
+147:  libX11 = super.libX11.overrideAttrs (attrs: {
+```
+
+### `nix-locate`
+
+Consider using `nix-locate` from the [`nix-index`](https://github.com/nix-community/nix-index) tool to find derivations that provide what you need.
+
 You will need the `Xlib.h` headers from the `X11` C package, the Nixpkgs derivation for which is `libX11`, available in the `xorg` package set.
```

Comment:
This conclusion should be common for all three subsections here. Maybe move it to the beginning of the section, something like

> You will need the `Xlib.h` headers from the `X11` C package, the Nixpkgs derivation for which is `libX11`, available in the `xorg` package set.
>
> There's different ways to figure this out: